GPU ported and optimised find_center_360: >420x speedup

team-gpu opened the pull request:
## Overview

This PR ports the `find_center_360` function to the GPU using cupy, and optimises the function on both CPU and GPU. It also adds appropriate tests and it uses a unified function name for both CPU and GPU.

## Performance (426x faster)

Based on the performance test in this PR, with normalisation, filtering, and overlap enabled. Measured on Diamond cluster, V100

* Starting performance CPU: **~3,070ms**
* Straight port to Cupy, using `get_array_module`: **~7,150ms**
* Final performance CPU: **~1,580ms**
* Final performance GPU: **~7.2ms** (426x faster than initial CPU, 993x faster than straight GPU port)

## Details

* The key item was to put the calculation of all the correlation metrics (with all possible position shifts) into a single kernel
* The Pearson correlation coefficient can in fact be calculated easily by hand, for every position in a different thread block
* Also removed a lot of temporary arrays and small operations, which also benefited the CPU implementation
* Apart from that, unified tests have been added and the CUDA code has been moved into its own folder, which makes development much easier and code better organised

namannimmo10 commented on httomolib/recon/rotation.py:
```diff
@@ -682,16 +780,17 @@ def _calculate_curvature(list_metric):
     min_pos : float
         Position of the minimum value with sub-pixel accuracy.
     """
+    xp = cp.get_array_module(list_metric)
```
why do you need xp here?

team-gpu commented:
Let me know if you have further questions / comments on this PR.

team-gpu commented:
I've refactored this PR as discussed - now the production code is GPU only, and the CPU code has been moved into the tests for reference purposes only. Note that the method name has been changed to add the suffix `_cupy` - not sure if that's desired?

namannimmo10 commented on tests/test_recon/test_recon.py:
```diff
@@ -31,19 +32,44 @@ def test_find_center_vo_cupy_ones(ensure_clean_memory):
 
+@cp.testing.gpu
+@pytest.mark.perf
+def test_find_center_vo_cupy_performance():
```
We already have this in `master` -- rebase with it, please.

team-gpu replied:
I have rebased it now - the reason this shows is that I moved it higher up - from the bottom of the file to this place. That's to keep the `vo_cupy` and the `360`  tests together.

dkazanc commented:
> I've refactored this PR as discussed - now the production code is GPU only, and the CPU code has been moved into the tests for reference purposes only. Note that the method name has been changed to add the suffix `_cupy` - not sure if that's desired?

Thanks! I think the suffix `cupy` is now redundant and can be dropped for all methods.

team-gpu commented:
> > I've refactored this PR as discussed - now the production code is GPU only, and the CPU code has been moved into the tests for reference purposes only. Note that the method name has been changed to add the suffix `_cupy` - not sure if that's desired?
> 
> Thanks! I think the suffix `cupy` is now redundant and can be dropped for all methods.

I removed it for this PR. We'll need to do a global refactoring to make all methods and tests consistent with this strategy after.
